### PR TITLE
New package: py-nanobind

### PR DESCRIPTION
**Pull request by chrisrichardson**
This is a new python/C++ interface library, which is lighter weight than the popular `py-pybind11` library. We wish to use it for our `py-fenics-basix` package, once available in spack.

### CODE REVIEW

**Comment by ma595** [comment]
@adamjstewart Does this look good to you?

File: var/spack/repos/builtin/packages/py-nanobind/package.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    version("1.2.0", tag="v1.2.0", submodules=True)
+
+    depends_on("python@3.8:", type=("build", "run"))
+    depends_on("py-pip", type="build")
+    depends_on("py-wheel", type="build")
+    depends_on("py-setuptools@42:", type="build")
+    depends_on("py-pytest", type="test")
+    depends_on("py-scikit-build", type="run")
+
+    depends_on("cmake@3.17:", type="run")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nanobind/package.py]: Technically this should be `py-cmake`. Also missing dep on `py-ninja`. Both are required at build and run time.

File: var/spack/repos/builtin/packages/py-nanobind/package.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    git = "https://github.com/wjakob/nanobind.git"
+
+    maintainers("ma595")
+
+    version("master", branch="master", submodules=True)
+    version("1.2.0", tag="v1.2.0", submodules=True)
+
+    depends_on("python@3.8:", type=("build", "run"))
+    depends_on("py-pip", type="build")
+    depends_on("py-wheel", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nanobind/package.py]: These are automatically added by the base class so you can remove them

File: var/spack/repos/builtin/packages/py-nanobind/package.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    depends_on("py-setuptools@42:", type="build")
+    depends_on("py-pytest", type="test")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nanobind/package.py]: We don't currently run any tests other than importing all files, so pytest probably won't do anything

File: var/spack/repos/builtin/packages/py-nanobind/package.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    depends_on("py-scikit-build", type="run")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nanobind/package.py]:
```suggestion
    depends_on("py-scikit-build", type="build")
```

File: var/spack/repos/builtin/packages/py-nanobind/package.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    homepage = "https://nanobind.readthedocs.io"
+    url = "https://github.com/wjakob/nanobind/archive/refs/tags/v1.2.0.tar.gz"
+    git = "https://github.com/wjakob/nanobind.git"
+
+    maintainers("ma595")
+
+    version("master", branch="master", submodules=True)
+    version("1.2.0", tag="v1.2.0", submodules=True)
+
+    depends_on("python@3.8:", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nanobind/package.py:29]: Where do you see that 3.8+ is required?

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-nanobind/package.py (hunk 5 above)
Comment (garth-wells) [reply var/spack/repos/builtin/packages/py-nanobind/package.py:29]: See  https://nanobind.readthedocs.io/en/latest/#dependencies and https://github.com/wjakob/nanobind/blob/6f0c3feaf088e78c75f2abee90164f20446eba08/CMakeLists.txt#L60.

**Comment by chrisrichardson** [comment]
@ma595 - can you add the latest version? 1.4.0 I think

File: var/spack/repos/builtin/packages/py-nanobind/package.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    version("master", branch="master", submodules=True)
+    version("1.4.0", tag="v1.4.0", submodules=True)
+    version("1.2.0", tag="v1.2.0", submodules=True)
+
+    depends_on("python@3.8:", type=("build", "run"))
+    depends_on("py-setuptools@42:", type="build")
+    depends_on("py-scikit-build", type="build")
+
+    depends_on("py-cmake@3.17:")
+    depends_on("py-ninja")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nanobind/package.py]:
```suggestion
    depends_on("py-cmake@3.17:", type="build")
    depends_on("py-ninja", type="build")
```

**Review by adamjstewart — CHANGES_REQUESTED** [review]

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**